### PR TITLE
fix for availability in hidden sections

### PR DESCRIPTION
**Pull request by Syxton**
Possible fix for #130

### CODE REVIEW

**Comment by Syxton** [comment]
@PhMemmel if you have a moment I wouldn't mind a second pair of eyes before I merge this.  We were seeing this issue on 4.4+ and 4.5.

**Comment by PhMemmel** [comment]
Will have a look as soon as I find some time. Thanks for finding this and providing a fix!

File: classes/actions.php [hunk 1]
```diff
@@ -95,23 +95,32 @@ public static function set_visibility(array $modules, bool $visible, bool $visib
...
+                // In a visible section, show has a visibleonpage of 1 and stealth has a visibleonpage of 0.
+                // Gather more information about the module and where it is located.
+                $modinfo = get_fast_modinfo($cm->course);
+                $format = course_get_format($modinfo->get_course());
+                $cm = $modinfo->get_cm($cm->id);
 
-            $visibleint = $visible ? 1 : 0;
-            $visibleonpageint = $visibleonpage ? 1 : 0;
+                $allowstealth = $format->allow_stealth_module_visibility($cm, $cm->get_section_info());
+                $visibleonpageint = ($allowstealth) ? 0 : 1;
```
Comment (PhMemmel) [inline classes/actions.php:120]:
```suggestion
                $visibleonpageint = $allowstealth ? 0 : 1;
```

File: classes/actions.php [hunk 2]
```diff
@@ -95,23 +95,32 @@ public static function set_visibility(array $modules, bool $visible, bool $visib
         }
 
         foreach ($modules as $cm) {
+            // Convert boolean to integer.
+            $visibleint = $visible ? 1 : 0;
+            $visibleonpageint = $visibleonpage ? 1 : 0;
+
+            // Stealth activies, if enabled are stored differently in visible and hidden sections.
```
Comment (PhMemmel) [inline classes/actions.php:102]:
```suggestion
            // Stealth activities, if enabled, are stored differently in visible and hidden sections.
```

**Review by PhMemmel — COMMENTED** [review]
LGTM. This moodle feature is pretty confusing :)

I wonder if there is no moodle core function to handle this, so we do not have to rebuild the logic here, but I assume the answer is "no", otherwise we would already use it :)

I also tested this manually, works as intended, bug fixed.

I would suggest to add this case to the unit test function?

**Comment by Syxton** [comment]
@PhMemmel ok I have rebased this change to the 4.5 version.  I also reviewed the unit tests and it already tested correctly, however what we really needed was a behat test making sure that in all cases it was showing correctly to teachers and students.  That is now done.  However, in the case that somehow a hidden section module gets set to visible: 1 and visibleoncoursepage: 0, there is still the strange case that Moodle will select "Hidden from students" in the activity edit screen under Availability.  Seems like a Moodle bug.

The logic is (visible, visibleoncoursepage).  
Visible sections: Visible (1, 1), Hidden (0, ~), Avail (1, 0)   
Hidden sections: Hidden (0, ~), Avail (1, 1)

**Comment by PhMemmel** [comment]
I retested and to be honest the status under availability is correct for me? Or have I the wrong reproduction steps?

**Comment by Syxton** [comment]
@PhMemmel The steps were (using an older version of mass actions), Create a hidden course section.  Place an activity in the hidden section.  Use mass actions to make the activity in the hidden section available.  It should show available.  Edit the activity, and under Common module settings -> Availability it will show "Hide on course page".  Another way is to manually set the activity in the hidden section to visible = 1 and visibleincourse = 0.

**Comment by PhMemmel** [comment]
I followed your steps. With this branch I cannot reproduce this anymore. It properly shows "available, but don't show on course page" in the activity edit form. So I think it can't become better than the current state it. Ready to merge I believe?

**Pull request merged**